### PR TITLE
Fix regression of EVP_PKEY_CTX_add1_hkdf_info() with older providers

### LINKED ISSUE #24611: EVP_PKEY_CTX_add1_hkdf_info() no longer works with providers that do not expose gettable parameters in ossl_kdf_hkdf_keyexch_functions array
Upgrading from OpenSSL 3.0.13 to 3.0.14 has uncovered a breakage introduced by 4953ab1a when using a provider derived from OpenSSL 3.0.13 or earlier.

The newly introduced `evp_pkey_ctx_add1_octet_string` function (called by EVP_PKEY_CTX_add1_hkdf_info) in evp/pmeth_lib.c contains the following [logic](https://github.com/openssl/openssl/commit/4953ab1aefd14db7038e28d62c0e3efb22ddb199#diff-066dd8b9ed11d5a34d72f0249b22f85422e3177f11d0057708dc4c85b2bea5d5R1058):
```
   /* Get the original value length */
    os_params[0] = OSSL_PARAM_construct_octet_string(param, NULL, 0);
    os_params[1] = OSSL_PARAM_construct_end();

    if (!EVP_PKEY_CTX_get_params(ctx, os_params))
        return 0;

    /* Older provider that doesn't support getting this parameter */
    if (os_params[0].return_size == OSSL_PARAM_UNMODIFIED)
        return evp_pkey_ctx_set1_octet_string(ctx, fallback, param, op, ctrl, data, datalen);
```

However, EVP_PKEY_CTX_get_params() will return 0 if `ctx->op.kex.exchange->get_ctx_params` is NULL and thus the subsequent check for older providers above does not get reached in the case where the provider does not expose a get_ctx_params function at all.

Any provider derived from older versions of OpenSSL will not expose get_ctx_params/gettable_params in the `ossl_kdf_*_keyexch_functions` array for key derivation algorithms. The commit referenced above caused these functions to be exposed: https://github.com/openssl/openssl/commit/4953ab1aefd14db7038e28d62c0e3efb22ddb199#diff-85fe36a1da6d2470d1a9ceb431d681dc458e600b994bb8df362e4564d943c698R246, so 3.0.14 is self-consistent.

We have worked around this locally by adjusting the code in 3.0.14 to check if any parameters are gettable when `EVP_PKEY_CTX_get_params` returns 0 in the logic above -- if none are, we assume that we're dealing with an old provider that does not expose get_params/gettable_params at all and fall back to `evp_pkey_ctx_set1_octet_string` in the same way as above.

### PR DESCRIPTION
**Pull request by t8m**
If there is no get_ctx_params() implemented in the key exchange provider implementation the fallback will not work. Instead check the gettable_ctx_params() to see if the fallback should be performed.

Fixes #24611

### CODE REVIEW

**Comment by t8m** [comment]
Tests exempted because this is really a corner case for old providers compatibility.

**Review by xnox — APPROVED** [review]
Cherrypick of this on top of 3.3.1 resolves runtime incompatibility with FIPS provider 3.0.9 for me.

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
ping @openssl/committers for the second review

**Review by tom-cosgrove-arm — APPROVED** [review]

**Pull request closed without merging**
